### ISSUE TITLE
informational: grinding leads to non-deterministic proof generation when multi-threaded

### ISSUE THREAD

**Issue by jonathanpwang** [Plonky3/Plonky3#556]
https://github.com/Plonky3/Plonky3/blob/9b891fa9ced28b9bb28bfe072f6c8f0042cf9f2c/challenger/src/grinding_challenger.rs#L35

Not considered a bug, but can lead to unexpected behavior: typically when blinding/zero-knowledge is "off" you would expect that proof generation is deterministic. However when `proof_work_bits != 0`, the grinding in fact will result in non-deterministic proofs when `parallel` is on due to concurrent grinding returning the first witness that satisfies the grinding condition.

**Comment by dlubarov** [comment]
Makes sense! Is the current behavior okay for you? If needed I figure we could avoid this with `find_first` instead of `find_any`. Might be some performance penalty but it might be negligible.

**Comment by jonathanpwang** [comment]
Yes the current behavior is okay, and it matches the behavior once a hiding pcs is used (which I saw you merged). It just took a bit to figure out where this came from, so I thought I'd share :)

**Comment by Nashtare** [comment]
I'll close this as any use of RNG leads to non-determinism, unless starting from a fixed seed (naturally not suited for production use).
